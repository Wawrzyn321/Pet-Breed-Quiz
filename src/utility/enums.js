const AppState = Object.freeze({
  Quiz: "QUIZ",
  Menu: "MENU"
});

const QuizState = Object.freeze({
  None: "NONE",
  ShowingQuestion: "SHOWING_QUESTION",
  ShowingAnswer: "SHOWING_ANSWER",
  Finished: "FINISHED"
});

const Categories = Object.freeze({
  Dogs: "DOGS",
  Cats: "CATS",
  Both: "BOTH",
  None: "NONE"
});

const Creatures = Object.freeze({
  Dog: "DOG",
  Cat: "CAT"
});

const AnswerState = Object.freeze({
  None: "NONE",
  Bad: "BAD",
  Good: "GOOD"
});

export { AppState, QuizState, Categories, Creatures, AnswerState };
