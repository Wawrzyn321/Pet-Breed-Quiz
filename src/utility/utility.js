const CONNECTION_TIMEOUT_MESSAGE = "Connection timeout";
const NETWORK_ERROR_MESSAGE = "NetworkError when attempting to fetch resource.";

function isConnectionError(error) {
  return (
    error.message === CONNECTION_TIMEOUT_MESSAGE ||
    error.message === NETWORK_ERROR_MESSAGE
  );
}

function getShuffledArray(sourceArray) {
  const array = [...sourceArray];
  for (let i = array.length - 1; i > 0; i--) {
    const j = getRandomNumber(i + 1);
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
}

function getRandomElements(array, count) {
  if (count > array.length) {
    throw new Error(
      "getRandomIndices: count must be less or equal the array length!"
    );
  }

  const shuffled = getShuffledArray(array);
  return shuffled.splice(0, count);
}

function getRandomNumber(maxExclusive) {
  return Math.floor(Math.random() * maxExclusive);
}

function fetchTimeout(url, options, timeout = 10000) {
  return Promise.race([
    fetch(url, options),
    new Promise((resolve, reject) =>
      setTimeout(() => reject(new Error(CONNECTION_TIMEOUT_MESSAGE)), timeout)
    )
  ]);
}

export {
  isConnectionError,
  getShuffledArray,
  getRandomElements,
  getRandomNumber,
  fetchTimeout
};
