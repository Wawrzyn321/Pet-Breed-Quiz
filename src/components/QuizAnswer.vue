<template>
  <li @click="markAnswer" :class="classObject">
    {{ answer.displayName }}
  </li>
</template>

<script>
import { AnswerState } from "./../utility/enums";

export default {
  computed: {
    classObject() {
      return {
        valid: this.answer.state === AnswerState.Good,
        invalid: this.answer.state === AnswerState.Bad
      };
    }
  },
  methods: {
    markAnswer() {
      this.$emit("answer-marked", this.answer);
    }
  },
  props: {
    answer: {
      required: true,
      type: Object
    }
  }
};
</script>

<style lang="scss" scoped>
li {
  cursor: pointer;
  font-size: 1.8em;
  list-style-type: none;
  border: 2px solid $text-dark;
  display: inline;
  padding: 3px;
  text-align: center;
  border-radius: 12px;
  text-transform: capitalize;
}

li:hover:not(.valid):not(.invalid) {
  background-color: $bg-color-hover;
  transition: 0.6s;
}

li.valid {
  background-color: $answer-valid;
  transition: 0.3s;
}

li.invalid {
  background-color: $answer-invalid;
  transition: 0.3s;
}
</style>
