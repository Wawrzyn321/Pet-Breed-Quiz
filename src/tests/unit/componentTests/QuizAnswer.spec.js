import QuizAnswer from "@/components/QuizAnswer";
import { shallowMount } from "@vue/test-utils";
import { AnswerState } from "@/utility/enums";

describe("QuizAnswer", () => {
  describe("Applying classes", () => {
    it("should not apply any classes when answer state is none", () => {
      const answer = {
        state: AnswerState.None,
        displayName: "Malamute",
        linkName: "malamute"
      };
      const wrapper = shallowMount(QuizAnswer, {
        propsData: { answer }
      });
      expect(wrapper.classes()).not.toContain("valid");
      expect(wrapper.classes()).not.toContain("invalid");
    });

    it("should apply valid classes if answer is marked good", () => {
      const answer = {
        state: AnswerState.Good,
        displayName: "Malamute",
        linkName: "malamute"
      };
      const wrapper = shallowMount(QuizAnswer, {
        propsData: { answer }
      });
      expect(wrapper.classes()).toContain("valid");
      expect(wrapper.classes()).not.toContain("invalid");
    });

    it("should apply valid classes if answer is marked bad", () => {
      const answer = {
        state: AnswerState.Bad,
        displayName: "Malamute",
        linkName: "malamute"
      };
      const wrapper = shallowMount(QuizAnswer, {
        propsData: { answer }
      });
      expect(wrapper.classes()).toContain("invalid");
      expect(wrapper.classes()).not.toContain("valid");
    });
  });

  it("should display answer's displayName", () => {
    const answer = {
      state: AnswerState.None,
      displayName: "Malamute",
      linkName: "malamute"
    };
    const wrapper = shallowMount(QuizAnswer, {
      propsData: { answer }
    });
    expect(wrapper.text()).toBe(answer.displayName);
  });

  it("should emit answer-marked event on clicking answer", () => {
    const wrapper = shallowMount(QuizAnswer, {
      propsData: { answer: { displayName: "answer" } }
    });

    wrapper.find("li").trigger("click");

    expect(wrapper.emitted()).toEqual({
      "answer-marked": [[{ displayName: "answer" }]]
    });
  });
});
