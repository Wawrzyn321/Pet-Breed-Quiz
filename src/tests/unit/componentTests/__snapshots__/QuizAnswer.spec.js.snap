// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizAnswer renders with minimal props 1`] = `
<li
  class=""
>
  
  Malamute

</li>
`;
